// @nguniversal/common/bundles/common-engine.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('fs'), require('tslib'), require('@angular/compiler'), require('@angular/core'), require('@angular/platform-server')) :
    typeof define === 'function' && define.amd ? define('@nguniversal/common/engine', ['exports', 'fs', 'tslib', '@angular/compiler', '@angular/core', '@angular/platform-server'], factory) :
    (global = global || self, factory((global.nguniversal = global.nguniversal || {}, global.nguniversal.common = global.nguniversal.common || {}, global.nguniversal.common.engine = {}), global.fs, global.tslib, global.ng.compiler, global.ng.core, global.ng.platformServer));
}(this, function (exports, fs, tslib_1, compiler, core, platformServer) { 'use strict';

    /**
     * @license
     * Copyright Google LLC All Rights Reserved.
     *
     * Use of this source code is governed by an MIT-style license that can be
     * found in the LICENSE file at https://angular.io/license
     */
    /** ResourceLoader implementation for loading files */
    var FileLoader = /** @class */ (function () {
        function FileLoader() {
        }
        FileLoader.prototype.get = function (url) {
            return new Promise(function (resolve, reject) {
                fs.readFile(url, function (err, buffer) {
                    if (err) {
                        return reject(err);
                    }
                    resolve(buffer.toString());
                });
            });
        };
        return FileLoader;
    }());

    /**
     * A common rendering engine utility. This abstracts the logic
     * for handling the platformServer compiler, the module cache, and
     * the document loader
     */
    var CommonEngine = /** @class */ (function () {
        function CommonEngine(moduleOrFactory, providers) {
            if (providers === void 0) { providers = []; }
            this.moduleOrFactory = moduleOrFactory;
            this.providers = providers;
            this.factoryCacheMap = new Map();
            this.templateCache = {};
        }
        /** Return an instance of the platformServer compiler */
        CommonEngine.prototype.getCompiler = function () {
            var compilerFactory = platformServer.platformDynamicServer().injector.get(core.CompilerFactory);
            return compilerFactory.createCompiler([
                { providers: [{ provide: compiler.ResourceLoader, useClass: FileLoader, deps: [] }] }
            ]);
        };
        /**
         * Render an HTML document for a specific URL with specified
         * render options
         */
        CommonEngine.prototype.render = function (opts) {
            return tslib_1.__awaiter(this, void 0, void 0, function () {
                var doc, _a, extraProviders, factory;
                return tslib_1.__generator(this, function (_b) {
                    switch (_b.label) {
                        case 0:
                            _a = opts.document;
                            if (_a) return [3 /*break*/, 2];
                            return [4 /*yield*/, this.getDocument(opts.documentFilePath)];
                        case 1:
                            _a = (_b.sent());
                            _b.label = 2;
                        case 2:
                            doc = _a;
                            extraProviders = tslib_1.__spread((opts.providers || []), (this.providers || []), [
                                {
                                    provide: platformServer.INITIAL_CONFIG,
                                    useValue: {
                                        document: doc,
                                        url: opts.url
                                    }
                                }
                            ]);
                            return [4 /*yield*/, this.getFactory()];
                        case 3:
                            factory = _b.sent();
                            return [2 /*return*/, platformServer.renderModuleFactory(factory, { extraProviders: extraProviders })];
                    }
                });
            });
        };
        /** Return the factory for a given engine instance */
        CommonEngine.prototype.getFactory = function () {
            var _this = this;
            // If module has been compiled AoT
            var moduleOrFactory = this.moduleOrFactory;
            if (moduleOrFactory instanceof core.NgModuleFactory) {
                return Promise.resolve(moduleOrFactory);
            }
            else {
                // we're in JIT mode
                var moduleFactory = this.factoryCacheMap.get(moduleOrFactory);
                // If module factory is cached
                if (moduleFactory) {
                    return Promise.resolve(moduleFactory);
                }
                // Compile the module and cache it
                return this.getCompiler().compileModuleAsync(moduleOrFactory)
                    .then(function (factory) {
                    _this.factoryCacheMap.set(moduleOrFactory, factory);
                    return factory;
                });
            }
        };
        /** Retrieve the document from the cache or the filesystem */
        CommonEngine.prototype.getDocument = function (filePath) {
            var doc = this.templateCache[filePath] = this.templateCache[filePath] ||
                fs.readFileSync(filePath).toString();
            // As  promise so we can change the API later without breaking
            return Promise.resolve(doc);
        };
        return CommonEngine;
    }());

    /**
     * @license
     * Copyright Google LLC All Rights Reserved.
     *
     * Use of this source code is governed by an MIT-style license that can be
     * found in the LICENSE file at https://angular.io/license
     */

    /**
     * @license
     * Copyright Google LLC All Rights Reserved.
     *
     * Use of this source code is governed by an MIT-style license that can be
     * found in the LICENSE file at https://angular.io/license
     */

    /**
     * @license
     * Copyright Google LLC All Rights Reserved.
     *
     * Use of this source code is governed by an MIT-style license that can be
     * found in the LICENSE file at https://angular.io/license
     */

    /**
     * Generated bundle index. Do not edit.
     */

    exports.ɵFileLoader = FileLoader;
    exports.ɵCommonEngine = CommonEngine;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
