// @ngx-lite/json-ld/esm5/ngx-json-ld.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Component, HostBinding, Input, ChangeDetectionStrategy } from '@angular/core';
import { DomSanitizer } from '@angular/platform-browser';
var NgxJsonLdComponent = /** @class */ (function () {
    function NgxJsonLdComponent(sanitizer) {
        this.sanitizer = sanitizer;
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    NgxJsonLdComponent.prototype.ngOnChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
        this.jsonLD = this.getSafeHTML(changes["json"].currentValue);
    };
    /**
     * @param {?} value
     * @return {?}
     */
    NgxJsonLdComponent.prototype.getSafeHTML = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        /** @type {?} */
        var json = value
            ? JSON.stringify(value, null, 2).replace(/<\/script>/g, '<\\/script>')
            : '';
        /** @type {?} */
        var html = "<script type=\"application/ld+json\">" + json + "</script>";
        return this.sanitizer.bypassSecurityTrustHtml(html);
    };
    NgxJsonLdComponent.decorators = [
        { type: Component, args: [{
                    selector: 'ngx-json-ld',
                    template: '',
                    changeDetection: ChangeDetectionStrategy.OnPush
                }] }
    ];
    /** @nocollapse */
    NgxJsonLdComponent.ctorParameters = function () { return [
        { type: DomSanitizer }
    ]; };
    NgxJsonLdComponent.propDecorators = {
        json: [{ type: Input }],
        jsonLD: [{ type: HostBinding, args: ['innerHTML',] }]
    };
    return NgxJsonLdComponent;
}());
export { NgxJsonLdComponent };
if (false) {
    /** @type {?} */
    NgxJsonLdComponent.prototype.json;
    /** @type {?} */
    NgxJsonLdComponent.prototype.jsonLD;
    /** @type {?} */
    NgxJsonLdComponent.prototype.sanitizer;
}
